File: MPPCS/Repository/SaleDatabaseRepository.cs
using CommonDomain;
using Persistence.Repository.Utils;
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Persistence.Repository
{
    public class SaleDatabaseRepository : SaleRepository
    {
        private DbUtils dbUtils;
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        public SaleDatabaseRepository(string url)
        {
            dbUtils = new DbUtils(url);
        }

        public void Add(Sale sale)
        {
            log.Debug("adding " + sale.GetName());
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "INSERT INTO Sales(buyer_name,tickets_bought,show_id) VALUES(@1,@2,@3)";
                command.Parameters.Add(new SQLiteParameter("@1", sale.GetName()));
                command.Parameters.Add(new SQLiteParameter("@2", sale.GetTicketsBought()));
                command.Parameters.Add(new SQLiteParameter("@3", sale.GetShowId()));

                int result = command.ExecuteNonQuery();
                log.Debug("SAVED: " + result);
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
        }

        public Sale Delete(int id)
        {
            log.Debug("deleting" + id);
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "DELETE FROM Sales WHERE id = @1";
                command.Parameters.Add(new SQLiteParameter("@1", id));
                Sale deleted = Find(id);
                int result = command.ExecuteNonQuery();
                log.Debug("DELETED: " + result);
                return deleted;
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
        }

        public Sale Find(int id)
        {
            log.Debug("finding " + id);
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "SELECT id,buyer_name,tickets_bought,show_id FROM Sales WHERE id = @1";
                command.Parameters.Add(new SQLiteParameter("@1", id));
                SQLiteDataReader dataReader = command.ExecuteReader();
                dataReader.Read();
                Sale sale = new Sale(
                    dataReader.GetString(1),
                    dataReader.GetInt32(2),
                    dataReader.GetInt32(3));
                sale.SetId(dataReader.GetInt32(0));
                dataReader.Close();
                log.Debug("FOUND " + sale.GetId());
                return sale;
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
        }

        public List<Sale> GetAll()
        {
            log.Debug("GetAll");
            List<Sale> sales = new List<Sale>();
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "SELECT * FROM Sales";
                SQLiteDataReader dataReader = command.ExecuteReader();

                while (dataReader.Read())
                {
                    Sale sale = new Sale(
                        dataReader.GetString(1),
                        dataReader.GetInt32(2),
                        dataReader.GetInt32(3));
                    sale.SetId(dataReader.GetInt32(0));
                    sales.Add(sale);
                }
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
            log.Debug("GOT ALL " + sales.Count());
            return sales;
        }

        public int Size()
        {
            log.Debug("size");
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "SELECT COUNT(*) AS count FROM Sales";
                SQLiteDataReader dataReader = command.ExecuteReader();
                dataReader.Read();
                int size = dataReader.GetInt32(0);
                dataReader.Close();
                log.Debug("SIZE IS " + size);
                return size;
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
        }

        public Sale Update(Sale sale)
        {
            log.Debug("updating " + sale.GetId());
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "UPDATE Sales SET buyer_name=@1, tickets_bought=@2,show_id=@3 WHERE id=@4";
                command.Parameters.Add(new SQLiteParameter("@1", sale.GetName()));
                command.Parameters.Add(new SQLiteParameter("@2", sale.GetTicketsBought()));
                command.Parameters.Add(new SQLiteParameter("@3", sale.GetShowId()));
                command.Parameters.Add(new SQLiteParameter("@4", sale.GetId()));
                Sale updated = Find(sale.GetId());
                int result = command.ExecuteNonQuery();
                log.Debug("UPDATED " + result);
                return updated;
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
        }

        public List<Sale> FindSalesForShow(int id)
        {
            log.Debug("finding sales for show " + id);
            List<Sale> sales = new List<Sale>();
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "SELECT id,buyer_name,tickets_bought,show_id FROM Sales WHERE show_id=@1";
                command.Parameters.Add(new SQLiteParameter("@1", id));
                SQLiteDataReader dataReader = command.ExecuteReader();

                while (dataReader.Read())
                {
                    Sale sale = new Sale(
                        dataReader.GetString(1),
                        dataReader.GetInt32(2),
                        dataReader.GetInt32(3));
                    sale.SetId(dataReader.GetInt32(0));
                    sales.Add(sale);
                }
                return sales;
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException(ex.Message);
            }
        }
    }
}

File: PersistenceAndServices/src/main/java/repository/SaleDatabaseRepository.java
package repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import domain.Sale;
import repository.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class SaleDatabaseRepository implements SaleRepository {
    private JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public SaleDatabaseRepository(Properties props) {
        logger.info("Initialized SaleDatabaseRepository with properties: {}",props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public Sale find(Integer id) throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("SELECT id,buyer_name,tickets_bought,show_id FROM Sales WHERE id = ?");
            statement.setInt(1,id);
             ResultSet result = statement.executeQuery();
             result.next();
             int found_id = result.getInt("id");
             String buyerName = result.getString("buyer_name");
             int ticketsBought = result.getInt("tickets_bought");
             int show_id = result.getInt("show_id");
             Sale sale = new Sale(buyerName,ticketsBought,show_id);
             sale.setID(found_id);
             result.close();
             logger.traceExit(sale);
             return sale;
        }
        catch (SQLException ex) {
            logger.error(ex);
            throw new Exception("No such ID exists!");
        }
    }

    @Override
    public List<Sale> getAll() throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Sale> sales = new ArrayList<>();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("SELECT id,buyer_name,tickets_bought,show_id FROM Sales");
            ResultSet result = statement.executeQuery();
            while (result.next()) {
                int id = result.getInt("id");
                String buyerName = result.getString("buyer_name");
                int ticketsBought = result.getInt("tickets_bought");
                int show_id = result.getInt("show_id");
                Sale sale = new Sale(buyerName, ticketsBought, show_id);
                sale.setID(id);
                sales.add(sale);
            }
            result.close();
        }
        catch (SQLException ex) {
            logger.error(ex);
            throw new Exception("Error getting all sales!");
        }
        logger.traceExit(sales);
        return sales;
    }

    @Override
    public void add(Sale sale) throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("INSERT INTO Sales(buyer_name,tickets_bought,show_id) VALUES(?,?,?)");
            statement.setString(1,sale.getBuyerName());
            statement.setInt(2,sale.getTicketsBought());
            statement.setInt(3,sale.getShowID());
            int result = statement.executeUpdate();
            logger.trace("Saved {} instances",result);
        }
        catch (SQLException ex) {
            logger.error(ex);
            throw new Exception("Adding sale failed!");
        }
    }

    @Override
    public Sale delete(Integer id) throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("DELETE FROM Sales WHERE id = ?");
            statement.setInt(1,id);
            Sale deleted = find(id);
            int result = statement.executeUpdate();
            logger.trace("Deleted {} instances",result);
            logger.traceExit(deleted);
            return deleted;
        }
        catch (Exception ex) {
            logger.error(ex);
            throw new Exception("No such ID exists!");
        }
    }

    @Override
    public Sale update(Sale sale) throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("UPDATE Sales SET buyer_name=?, tickets_bought=?,show_id=? WHERE id=?");
            statement.setString(1,sale.getBuyerName());
            statement.setInt(2,sale.getTicketsBought());
            statement.setInt(3,sale.getShowID());
            statement.setInt(4,sale.getID());
            Sale updated = find(sale.getID());
            int result = statement.executeUpdate();
            logger.trace("Updated {} instances",result);
            logger.traceExit(updated);
            return updated;
        }
        catch (Exception ex) {
            logger.error(ex);
            throw new Exception("Updating sale failed!");
        }
    }

    @Override
    public Integer size() throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("SELECT COUNT(*) AS count FROM Sales");
            ResultSet result = statement.executeQuery();
            result.next();
            int size = result.getInt("count");
            result.close();
            logger.traceExit(size);
            return size;
        }
        catch (Exception ex) {
            logger.error(ex);
            throw new Exception("Error in database!");
        }
    }

    @Override
    public List<Sale> findSalesForShow(Integer showID) throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Sale> sales = new ArrayList<>();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("SELECT id,buyer_name,tickets_bought,show_id FROM Sales WHERE show_id=?");
            statement.setInt(1,showID);
            ResultSet result = statement.executeQuery();
            while (result.next()) {
                int id = result.getInt("id");
                String buyerName = result.getString("buyer_name");
                int ticketsBought = result.getInt("tickets_bought");
                int show_id = result.getInt("show_id");
                Sale sale = new Sale(buyerName, ticketsBought, show_id);
                sale.setID(id);
                sales.add(sale);
            }
            result.close();
        }
        catch (SQLException ex) {
            logger.error(ex);
            throw new Exception("Error getting all sales!");
        }
        logger.traceExit(sales);
        return sales;
    }
}
